File: kangaroo_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/frequentBuyer/FrequentBuyerHandler.kt
package com.kangaroo.flutterplugin.android.features.frequentBuyer

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kangaroorewards.appsdk.features.frequentBuyer.FrequentBuyerApi
import kangaroorewards.appsdk.features.frequentBuyer.serializeFrequentBuyerState


class FrequentBuyerHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_frequent_buyer"

    override val eventChannel: String
        get() = "customer_sdk/events/get_frequent_buyer"

    override fun onMethodCall(call: MethodCall): Unit? {
        return getFrequentBuyer(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        fun getFrequentBuyer(call: MethodCall): Unit? {
            FrequentBuyerApi().getFrequentBuyer()
            return null
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        FrequentBuyerApi().frequentBuyerState.serializeFrequentBuyerState().observe { result ->
            pushSerializedResultToEventSink<FrequentBuyerApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_app_sdk_macos/macos/Classes/features/frequentBuyer/FrequentBuyerHandler.swift
import Foundation
import Flutter
import KangarooAppSDKiOS

class FrequentBuyerHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_frequent_buyer"

    var eventChannel: String = "customer_sdk/events/get_frequent_buyer"

    func onMethodCall(call: FlutterMethodCall) -> Void? {
        FrequentBuyerHandler.getFrequentBuyer(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getFrequentBuyer(call: FlutterMethodCall) {
        FrequentBuyerApi().getFrequentBuyer()

        
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        FrequentBuyerApi().frequentBuyerState
            .serializeFrequentBuyerState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
